packages/malloy/src/api/util.ts

```diff
@@ -0,0 +1,208 @@
+    // so here is a comment, for that day.
+    let parsed = DateTime.fromISO(value, {zone: 'UTC'});
+    if (!parsed.isValid) {
+      parsed = DateTime.fromSQL(value, {zone: 'UTC'});
+    }
+    return parsed.toJSDate();
+  }
+}
+
+export function mapData(data: QueryData, schema: Malloy.Schema): Malloy.Data {
+  function mapValue(
+    value: QueryValue,
+    field: Malloy.DimensionInfo
+  ): Malloy.Cell {
+    if (value === null) {
+      return {kind: 'null_cell'};
+    } else if (
+      field.type.kind === 'date_type' ||
+      field.type.kind === 'timestamp_type'
+    ) {
+      const time_value = valueToDate(value).toISOString();
+      if (field.type.kind === 'date_type') {
+        return {kind: 'date_cell', date_value: time_value};
+      } else {
+        return {kind: 'timestamp_cell', timestamp_value: time_value};
+      }
+    } else if (field.type.kind === 'boolean_type') {
+      if (typeof value !== 'boolean') {
+        throw new Error('Invalid boolean');
+      }
+      return {kind: 'boolean_cell', boolean_value: value};
+    } else if (field.type.kind === 'number_type') {
+      if (typeof value !== 'number') {
+        throw new Error('Invalid number');
+      }
+      return {kind: 'number_cell', number_value: value};
+    } else if (field.type.kind === 'string_type') {
+      if (typeof value !== 'string') {
+        throw new Error('Invalid string');
+      }
+      return {kind: 'string_cell', string_value: value};
+    } else if (field.type.kind === 'array_type') {
+      if (!Array.isArray(value)) {
+        throw new Error('Invalid array');
+      }
+      return {
+        kind: 'array_cell',
+        array_value: value.map(value =>
+          mapValue(value, {
+            name: 'foo',
+            type: (field.type as Malloy.AtomicTypeWithArrayType).element_type,
+          })
+        ),
+      };
+    } else if (field.type.kind === 'json_type') {
+      return {kind: 'json_cell', json_value: JSON.stringify(value)};
+    } else if (field.type.kind === 'sql_native_type') {
+      return {kind: 'sql_native_cell', sql_native_value: JSON.stringify(value)};
+    } else {
+      const type = field.type;
+      if (type.kind !== 'record_type') {
+        throw new Error(
+          `Invalid record in result ${JSON.stringify(field)}, ${JSON.stringify(
+            value
+          )}`
+        );
+      }
+      return mapRow(value as QueryDataRow, {
+        kind: 'join',
+        relationship: 'many',
+        name: 'foo',
+        schema: {
+          fields: type.fields.map(
+            f => ({kind: 'dimension', ...f}) as Malloy.FieldInfoWithDimension
+          ),
+        },
+      });
+    }
+  }
+  function mapRow(
+    row: QueryDataRow,
+    field: Malloy.FieldInfoWithJoin
+  ): Malloy.Cell {
+    const cells: Malloy.Cell[] = [];
+    for (const f of field.schema.fields) {
+      const value = row[f.name]; // TODO this might not work for weird names?
```

Comment:
TODO

Reply:
It's unclear whether DBs always return field names that exactly match the field names in Malloy...